src/transform/classify.rs

```diff
@@ -0,0 +1,64 @@
+/// Transforms for classifying (tagging and categorizing) events.
+///
+/// Based on code in aw_research: https://github.com/ActivityWatch/aw-research/blob/master/aw_research/classify.py
+
+use regex::Regex;
+use serde_json;
+
+use std::collections::HashSet;
+use crate::models::Event;
+
+pub fn classify(events: Vec<Event>, classes: Vec<(String, Regex)>) -> Vec<Event> {
+    // TODO: There is probably a better way that avoids the clone?
+    events.iter().map(|e| classify_one(e.clone(), &classes)).collect()
+}
+
+fn classify_one(mut event: Event, classes: &Vec<(String, Regex)>) -> Event {
+    let mut tags: HashSet<String> = HashSet::new();
+    for (cls, re) in classes {
+        for val in event.data.values() {
+            // TODO: Recurse if value is object/array
+            if val.is_string() && re.is_match(val.as_str().unwrap()) {
+                tags.insert(cls.clone());
+                break;
+            }
+        }
+    }
+
+    // An event can have many tags
+    event.data.insert("$tags".into(), serde_json::json!(tags));
+
+    // An event can only have one category, although the category may have a hierarchy,
+    // for instance: "Work -> ActivityWatch -> aw-server-rust"
+    // A category is chosed out of the tags used some rule (such as picking the one that's deepest in the hierarchy)
+    let category = choose_category(tags);
+    event.data.insert("$category".into(), serde_json::json!(category));
+    event
+}
+
+fn choose_category(tags: HashSet<String>) -> String {
+    tags.iter().fold(&"Uncategorized".to_string(), |acc: &String, item: &String| {
+        if item.matches("->").count() >= acc.matches("->").count() {
+            item
+        } else {
+            acc
+        }
+    }).clone()
+}
+
+#[test]
+fn test_classify() {
```

Comment:
Currently we have all tests in the tests folder.

The Rust way to do things is to have unittests inside the source and integration tests inside the tests directory. Not sure which way is the best way for us.

Reply:
I quite like having small unittests close to the function definition. Although it doesn't seem to show up in the coverage properly?

Reply:
Yeah, coverage is kind of broken currently anyway though, so whatever I guess